### DISCUSSION TITLE
Proposal: metrics_port in scrape config

### DISCUSSION THREAD

**Discussion by candlerb** [prometheus/prometheus#9466] [Ideas]
## Proposal

Add an optional `metrics_port` [scrape config](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#scrape_config) setting, which defaults to 0 (unset).

At scrape time, if the target `__address__` doesn't end with colon-port, then:
* If label `__metrics_port__` is set, use `__metrics_port__` as the port number
* Else if job-level `metrics_port` is set, use `metrics_port`
* Else do nothing, i.e. use `__address__` as-is without any port

This is done *after* `__address__` is copied to `instance` (which of course only happens if `instance` is unset)

**Use case. Why is this important?**

Currently, scheme and path can be set globally for a job, but port cannot.

For newcomers to prometheus: they do the "obvious" thing which is to create targets files like this:

```
# targets for job node_exporter
- host1.example.com:9100
- host2.example.com:9100
```

This results in two problems:
* they see messy labels in their dashboards
* when they get to more advanced queries, and want to join metrics from two different exporters, they find there is no matching `instance` label between them (because they embed two different port numbers)

They then have to learn and deploy relabelling tricks for [controlling the instance label](https://www.robustperception.io/controlling-the-instance-label), and modify their existing dashboards appropriately.  If the instance label had just been the hostname in the first place, this wouldn't be necessary.

Experienced prometheus users, who know about controlling instance labels, end up writing verbiage in every scrape job just to eliminate the port number:

```
    relabel_configs:
      - source_labels: [__address__]
        target_label: instance
      - source_labels: [__address__]
        target_label: __address__
        replacement: '${1}:9100'
```
 it would be so much simpler not to have to do this, and just set `metrics_port: 9100` instead.

**Backwards compatibility**

This is entirely(*) backwards compatible.  If you don't set `metrics_port` or `__metrics_port__` then nothing changes.  Even if you do, `__address__` values containing port numbers continue to be permitted and will take precedence.

I would expect that after a period of time, tutorials and examples would be updated to make use of `metrics_port` and remove the port numbers from targets files.

(*) in theory, someone might have a relabelling config which uses `__metrics_port__` as an internal temporary variable; but the documentation says they should use labels starting `__tmp` for this.

**Safety**

I do not believe that allowing the port number to be dropped from the `__address__` (and hence the default instance label) introduces any problems.  Firstly, this is good practice anyway.  Secondly, if `metrics_port` is set, this is at the job level.  Every job has a unique `job="<jobname>"` label, so two different scrapes with different `metrics_port` will have distinct label sets.

If you override using the `__metrics_port__` label then you can make collisions, but the same can be done with relabelling today.  I suspect that `__metrics_port__` will very rarely be used, in the same way as `__scheme__` and `__metrics_path__`, and is there really for completeness only.   Given that the port can still be set in `__address__`, then I wouldn't object strongly if `__metrics_path__` were left out.

**Bikeshed**

`port` rather than `metrics_port` ??

Many sd_configs have `port:` already, ~~which of course is a different port (for talking to the sd mechanism, rather than scraping the target).~~

~~Although the scrape_config could also gain a `port` (like `scheme`), on balance I think it is clearer to call it something else - for the same reason that `metrics_path` is not simply `path`.~~

**Comment by roidelapluie** [comment]
> Many sd_configs have port: already, which of course is a different port (for talking to the sd mechanism, rather than scraping the target).

This is incorrect, when SD's have this, it is to append to the targets for scraping.

**Comment by candlerb** [comment]
Many thanks for the correction.  The SD documentation does indeed say this is the port for metric scraping (with the exception of Triton which says "The port to use for discovery _and_ metric scraping")

I couldn't find it stated explicitly in the documentation, but it turns out that the service discovery mechanisms all set an `__address__` in addition to `__meta_xxx`; I wrongly thought you had to do your own relabelling to set `__address__`.

It was [ec2_sd_config](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#ec2_sd_config) which confused me.  How does it know whether you want to use `__meta_ec2_public_ip` or `__meta_ec2_private_ip` for scraping?  The answer is in the [source code](https://github.com/prometheus/prometheus/blob/v2.29.2/discovery/aws/ec2.go#L257-L258):

```
addr := net.JoinHostPort(*inst.PrivateIpAddress, fmt.Sprintf("%d", d.cfg.Port))
labels[model.AddressLabel] = model.LabelValue(addr)
```

That is: it's hard-coded that `__address__` is set by default to the private IP + SD port.  If that's not what you want, then you have to do relabelling to override it.  There is only a hint of this in the documentation:

```
# The port to scrape metrics from. If using the public IP address, this must
# instead be specified in the relabeling rule.
[ port: <int> | default = 80 ]
```

Thanks again for putting me right.  Now I have to consider how this might interact with my proposal.

I note that almost all the sd mechanisms which have a "port:" setting, default to port 80.  The exceptions are:
* [dns_sd_config](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#dns_sd_config), which doesn't have any default. If it's unset, and the record type is A or AAAA (not SRV), then an error is raised.
* [triton_sd_config](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#triton_sd_config), which defaults to 9163.

To do this in a backwards-compatible way, we don't want the `__address__` to change suddenly from `foo:80` to `foo`; this would affect `instance` labels, and could break existing rewriting rules which expect to see `__address__` in the format `address:port`.

What I propose is:

* if the SD stanza is within a job that *doesn't* have `metrics_port` set, then behave exactly as today
* if the SD stanza is within a job that *does* have `metrics_port` set, but the SD `port:` is not set, then don't append any port to `__address__` (i.e. the default SD `port` changes from "80" or "9193" to "no port")
    * EDIT: DNS SRV records always append a port

At this point, it becomes more reasonable to suggest that the setting be called `port` at the job level as well.

In summary:
- if you set `port` at the job level, it's used for scraping when the `__address__` doesn't contain a port
- if you set `port` at the SD level, it's used to append a port to `__address__` (as today)
- if you don't set `port` at either job or SD level, then the SD appends its default port to `__address__` (as today)

People can then migrate over time.

**Comment by roidelapluie** [comment]
> if you set port at the SD level, it's used to append a port to __address__ (as today)


That is not correct. Some SD's only use that when they can not determine which port to use (e.g. docker).

**Comment by roidelapluie** [comment]
I think that this proposal over-complicates things.

**Comment by SuperQ** [comment]
Yes, I think this proposal as-is doesn't make the situation better.

**Comment by roidelapluie** [comment]
An alternative would be to introduce shortcut actions in relabeling: `relabeling_configs: [action: remove-instance-port]`

**Comment by candlerb** [comment]
> I think that this proposal over-complicates things

Are you referring to the backwards-compatibility approach I've suggested, or the ability to specify a target address without port in general?

It seems odd to me that I can specify `scheme` and `metrics_path` at job level, but not `metrics_port`.  If it had been that way from day one, so that `__address__` was permitted to carry the address only, then the usual case of scraping the same exporter on a fixed port would have been simpler.  You'd have been able to use `__metrics_port__` label to deal with the less common case of the same job talking to variable ports (just as you can use `__metrics_path__` for the unusual case of the same job talking to different paths).  But instance labels would have been more useful.

Another option is to go further, and say that `__address__` can contain *any or all* parts of the scrape URL:

```
[http://]  1.2.3.4  [:9100]  [/metrics]
```

In this case `scheme`, `metrics_port` and `metrics_path` are only applied when the corresponding pieces are missing.  It would make the special labels `__scheme__` and `__metrics_path__` redundant, since you could apply those to `__address__` either directly, or via relabelling.

**Comment by roidelapluie** [comment]
One of the main use cases of Prometheus is with kubernetes, where there are may pods on the same IP.

In general, it is common to have the address and port in the same parameter, like we do for our --web.listen-address.

I do not think we should start adding preferences and order here.

**Comment by roidelapluie** [comment]
The rough consensus seems to go towards not accepting this idea.

What do you all think about the "shortcut" relabel action: [action: remove-instance-port] ?

Would it help addressing the initial issue @candlerb ?

**Comment by candlerb** [comment]
Not really.

The underlying problem is that the *default* instance label contains both address and port.  This makes it very hard to join metrics between two different exporters on the same node.

Once you've been bitten by this, you realise you have to throw away all your existing timeseries and do a whole load of relabelling, just to get "clean" instance labels.

I think what you're suggesting is something like this:

```
relabel_configs:
  - action: set_instance_without_port
```

as a magic shortcut for something like:

```
relabel_configs:
  - source_labels: [__address__]
    regex: '(.+):(\d+)'
    target_label: instance
    replacement: '${1}'
```

But whether you have to paste 2 extra lines into every scrape job or 5 extra lines doesn't really make much difference.  It's the fact that you have to understand *why* this change is required, and to avoid being bitten by it in the first place.

A secondary problem is that a ton of Grafana dashboards I find published out there are hard-coded to expect `host:port` in the instance label - which means you can't use them once you've cleaned up your instance labels, without major work (sometimes modifying every single PromQL query in the dashboard).

Now, I do realise there are cases where the port number is required to distinguish timeseries - for example if you are using docker SD (although that assumes that *all* the service ports which are exposed on a docker host are scrapable for metrics on the same path).  In such cases, I agree that either it would be necessary for the instance label to include the port, or for a separate label to be present with the port, just to make sure the time series are distinct.

But for cases like node_exporter and other node-level exporters, working at the level of the whole device or the whole VM, it would be really nice if the instance label were clean *by default*.

It also seems odd to me that "port" belongs in `__address__`, but "path" doesn't.  Taken to the other extreme, shouldn't `__address__` be the whole scrape URL, combining scheme, host, port and path?

If it has to be as it is, then so be it.  It's just something I need to train people on, and it means introducing relabelling as a concept earlier than I would otherwise like.

**Comment by roidelapluie** [comment]
Here is what I am doing: I am setting the `instance` label to the hostname while scraping via IP address. I do not like IP addresses in `instance` labels and I do not like to depend on DNS for monitoring. This is directly in the SD so Prometheus does not use the address here. So I do not have this issue (I mainly use file_sd).

**Comment by candlerb** [comment]
I do that too - and at that level of sophistication, relabelling *is* required to keep the targets file clean (*).

Simpler setups use the DNS name as both instance label and target; and others may be happy to see the IP address as both instance and target.  For those, it's either the DNS name or the IP address which identifies the target host.  The port number is just a scraping detail, rather like the metrics path: its presence in the *default* instance label hampers the usefulness of the metrics.

If you could set e.g. `metrics_port: 9100` at the scrape job level, and the individual targets were allowed to omit port numbers, then this issue wouldn't arise.

I guess what I'm suggesting could be implemented as an extra rule at the *end* of the relabel config, so let me try writing it as such:

```
relabel_configs:
  ... others (if any) go here

  # Duplicate the logic which prometheus already has
  - source_labels: [__address__, instance]
    regex: '(.*);'
    target_label: instance
  # Update the address
  - source_labels: [__address__]
    regex: '.*(:\d+)'
    target_label: __tmp_port
  - source_labels: [__address__, __tmp_port]
    regex: '(.*);'
    target_label: __address__
    replacement: '${1}:9100'
```

Updating the address a single rule is a bit awkward given that IPv6 addresses can contain colons and RE2 doesn't do assertions, but maybe something like this will work (untested):

```
...
  # Update the address
  - source_labels: [__address__]
    regex: '([^:]*|\\[.*\\])'
    target_label: __address__
    replacement: '${1}:9100'
```

----

(*) although in principle you could do it without relabelling, using an ugly file sd with one target per "targets" block:

```
- labels: { instance: foo }
  targets: [192.0.2.1:9100]
- labels: { instance: bar }
  targets: [192.0.2.2:9100]
...
```
